streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -580,4 +562,118 @@ private String getBrokerSideConfigValue(final Config brokerSideTopicConfig,
+                        Optional.of(replicationFactor)
+                    ).configs(newTopicConfigs.get(topicName))
+                ).collect(Collectors.toSet());
+
+            log.info("Going to create internal topics: " + newTopics);
+            final CreateTopicsResult createTopicsResult = adminClient.createTopics(newTopics);
+
+            final Map<String, KafkaFuture<Void>> createResultForTopic = createTopicsResult.values();
+            while (!createResultForTopic.isEmpty()) {
+                for (final InternalTopicConfig topicConfig : topicConfigs.values()) {
+                    final String topicName = topicConfig.name();
+                    if (!createResultForTopic.containsKey(topicName)) {
+                        throw new IllegalStateException("Create topic results do not contain internal topic " + topicName
+                            + " to setup. " + BUG_ERROR_MESSAGE);
+                    }
+                    final KafkaFuture<Void> createResult = createResultForTopic.get(topicName);
+                    if (createResult.isDone()) {
+                        try {
+                            createResult.get();
+                            topicStillToCreate.remove(topicName);
+                        } catch (final ExecutionException executionException) {
+                            final Throwable cause = executionException.getCause();
+                            if (cause instanceof TopicExistsException) {
+                                log.info("Internal topic {} already exists. Topic is probably marked for deletion. " +
+                                    "Will retry to create this topic later (to let broker complete async delete operation first)",
+                                    topicName);
+                            } else if (cause instanceof TimeoutException) {
+                                log.info("Creating internal topic {} timed out.", topicName);
+                            } else {
+                                log.error("Unexpected error during creation of internal topic: ", cause);
+                                throw new StreamsException(
+                                    String.format("Could not create internal topic %s for the following reason: ", topicName),
+                                    cause
+                                );
+                            }
+                        } catch (final InterruptedException interruptedException) {
+                            throw new InterruptException(interruptedException);
+                        } finally {
+                            createResultForTopic.remove(topicName);
+                        }
+                    }
+                }
+
+                maybeThrowTimeoutException(
+                    Collections.singletonList(topicStillToCreate),
+                    deadline,
+                    String.format("Could not create internal topics within %d milliseconds. This can happen if the " +
```

Comment:
In this case I think we should include some error details here. In particular, the last seen error for each topic. I'm worried about cases where we try to create but the create times out but is eventually successful. We'd return an error back, but the user would have no way to know that setup failed because an internal topic already exists.

Reply:
This sounds like a really useful idea.